Overhaul Dodge workshop

jkwok91 opened the pull request:
Closes https://github.com/hackclub/hackclub/issues/522

This is a condensed rewrite of the existing Dodge workshop, with improvements suggested in #522.

jkwok91 commented:
Okay, making a list of things I messed up/have to fix:
- Flesh out Sharing section
- Add Table of Contents
- Add link to demo
- Add finished code somewhere
- Add updated code at the end of each step, maybe?
- Add gifs
- Add links to the documentation for all the stuff in the inspiration section / or demos

jkwok91 commented:
Todos from #522:
- ~~keep player from going off screen~~ (already did this in my rewrite)
- show game over screen saying "game over" and a button to reset
  - using text() might be a hint as to how to display score
  - maybe need to be more explicit though?

zachlatta commented:
Just submitted https://github.com/jkwok91/hackclub/pull/2 with some preliminary improvements. I have a call in 8 hours, but I'll do a more thorough readthrough with feedback tomorrow.

zachlatta commented:
Feedback:

#### Overall

Really great start. I think nearly all issues can be fixed by watching a beginner go through the workshop and fixing it accordingly.

A few quick overall points:

- I think the way the `draw()` function works needs a better introduction. It's not clear that the way stuff "moves" on the screen is by calling that function 30-60 times a second with tiny changes between each call. Perhaps an analogy to the post-it note flip books would be good for this?

  ![](http://assignments.ds106.us/files/2015/07/old_school.gif)
  
- I think the introduction needs an overhaul to better explain what p5.js and p5.play are. What is a library? What's the difference between them and "regular JavaScript"?
- I'd put spaces after commas in arguments (`func(arg1, arg2, arg3)` instead of `func(arg1,arg2,arg3)`). More readable to beginners.
- Cloud9 is giving me a bajillion errors. I know to ignore them, but I could easily see this being stressful/confusing to a beginner. It'd be nice to either find a way to disable errors entirely or get Cloud9 to recognize the inclusion of p5.js and p5.play.

  ![](https://i.imgur.com/xNSqKHo.png)
    
#### Section by Section

- **Part I. Setup**
  - It's still not clear what p5.js and p5.play are
  - It's unclear what script tags do
- **Part II. The External JS File**
  - **Set Up**
    - Have we introduced functions yet? The user may not know what these are.
    - It's unclear what setup() and draw() do
    - It's unclear what a canvas is
      - How is the canvas different from index.html?
    - It's unclear how `0, 0, 100` corresponds to a dark blue background
    - It's unclear what RGB is.
  - **Adding the Player**
    - Where do I put `var player`? How does `var player` create a new player?
    - I could see tons of people not adding `var player` to the top of their file.
    - Where do `width` and `height` come from when declaring the existence of the player?
    - It's unclear what "placeholder for an image" means
    - The introduction of the coordinate system could be improved. I might introduce the coordinate system before having the user write the code to add the player to the screen. A visual explaining the different coordinate system would be super helpful.
    - It should probably be explained that the color of the player is randomly generated by p5.play. When I was playing with the position of the player, I thought I was accidentally changing the player's color.
    - I'd set the player's `y` coordinate to `height-40` just so there's some visual separation from the bottom to show that the player isn't being cut-off at all.

      ![](https://i.imgur.com/CwjyqYg.png)

  - **Adding Motion**
    - Wording can be improved throughout this section
    - You're starting to hint at how movement works in games here, but I really think this should be introduced more explicitly and earlier in the tutorial. Perhaps around when we create the `draw()` function.
    - It's unclear where to write the code block given in this section
  - **Adding Controls**
    - Where do the `LEFT_ARROW` and `RIGHT_ARROW` variables come from?
    - Will this be the first time the user is seeing conditionals? What does the old Dodge tutorial assume?
    - It's unclear where to put the code in this section. I think it may also not be obvious to the user that they should delete their movement code from the last section.
  - **Keeping the Player within Bounds**
    - Woah, woah? What's happening in this code snippet? May be worth more explicitly breaking down what's happening, especially in `keyDown(RIGHT_ARROW) && player.position.x < width`.
      - You explain what's happening in a paragraph right under the code block, but I think many users will just skim over that. Maybe bullet points might make it more visually distinctive?
    - Unclear where to put the code in this section. Just saying that you should update your code from the last section with this should be enough.
    - Half of the player goes off the screen with the restriction we add

      ![](https://i.imgur.com/swYyopg.gif)

    - Ah, I see. Tricky tricky you. The explanation for what's going wrong can be improved.
      - Might be worth explaining why you put `(width-25)` in parentheses. Order of operations, yo.
  - **Adding the Enemy**
    - May be worth customizing the player's sprite image before doing this.
  - **Moving the Enemy**
    - Somewhat unclear where to put `enemy.position.y = enemy.position.y + 1;`
    - Modifying to code to make the enemy faster could be more explicit.
    - Explanation for how the conditional for resetting the enemy could be improved. Right now it's not very explicit what that code is exactly doing.
    - Woah, where did `random` come from? Maybe some explanation for this?
      - Oh, I see. Explanation is below code snippet. Maybe put this above? Threw me off when I first read it.
  - **Collision**
    - I'm starting to get the hang of where to put the code from your snippets, but a little more context on where to put the code would be helpful.
    - It's unclear what a function is/how to create one. I might leave the `gameOver` function out of this section entirely and do something else for the lose condition (perhaps flash the screen red?). Then we could do a more thorough overview of how to create the `gameOver` function in the next section.
      - Yeah, adding on to this, when `gameOver` is called, the JavaScript on the page crashes, potentially giving the user the illusion that `gameOver` is a function defined by p5.js or p5.play. I can very easily see this being confusing to the user.
  - **Game Over**
    - Very first description about p5.js/p5.play vs JavaScript can be improved.
    - I'd include all of the arguments when calling the `background` function in `gameOver`. I imagine the concept of partial arguments would be confusing to the user.
    - Description for how game states work is okay, wording can be improved though. The concept of a "flag" may be foreign to the user.
      - It may also be unclear what booleans are. I'll leave whether to explain that up to you.
    - The introduction of `isGameOver` can probably be refactored to the following (with an explanation of `return`:

      ```js
      var isGameOver;
      
      function setup() {
        ...
        isGameOver = false;
        ...
      }
      
      function draw() {
        if (isGameOver) {
          gameOver();
          return;
        }

        // Insert everything else
      }
      ```

  - **Restart Game**
    - Small wording improvements could probably be made here, no major complaints
  - **Error Handling**
    - This section could probably have a better name
  - **Spicing it up**
    - Imgur is probably blocked on school networks. You can use our proxy to access it. https://surrogate.hackedu.us/i.imgur.com/<img_id>.ext
    - The example image given is broken.
- **Part III. Code Cleanup**
  - Why declare and then later initialize the constants in `setup`? I think it'd be cleaner to initialize them right when they're declared.
  - It's unclear where to put the constant declarations
  - I personally wouldn't introduce declaring multiple variables with a single `var` statement yet. Could be confusing to beginners. I'd do the following instead.

    ```js
    var PLAYER_WIDTH;
    var PLAYER_HEIGHT;
    var ENEMY_WIDTH;
    var ENEMY_HEIGHT;
    ```
    
  - Introduction of what a constant is could be made more direct.
  - The image used in this example for the player is different than what's used previously.
  - The code example with the full refactored code in this section could be improved
    - No multiple variable declarations
    - What is the `%` symbol in the comment for?
    - Instead of just providing the full refactored code here, I'd walk the user through replacing all usages of the player and enemy's sizes with the appropriate constants. I'd also walk them through making other changes to make the code cleaner. Remember, this is probably their second or third exposure to JavaScript. We want to make things as clear as possible for them.
- **Part IV. Publishing and Sharing**
  - Before telling them to share their project on `#shipit`, I'd walk them through pushing it to their GitHub Pages site (which they set up in the Personal Website workshop). This should be as easy as telling them to run `git add --all`, `git commit -m "Create Dodge game"`, and then `git push`.
- **Part V. Hacking**
  - Suggestions for what they want to work on next are great. I'd format them as bullet points to make them more visually distinct.

jkwok91 commented:
responses to feedback

## overview

- draw() function - i don't really know how to introduce this better because i feel like the user will realize how things move when they reach the step with adding 1 to position. NEEDS BEGINNER TESTING

- explaining p5.js/p5.play -DONE

- spaces - ok

- hiding errors - idk, we could hide the entire gutter, which would hide line numbers, which i do not like.

## setup

- i think you fixed these issues

## external js file

### setup

- i am not gonna introduce functions. there has to be a workshop before this one if functions still haven't been introduced. this workshop is getting massive bloat as it is.

- i specified that `setup()` and `draw()` create visuals. didn't mention that before, so someone going into this blind would be like "what are these seemingly arbitrary things." but idk how to explain that `setup()` sets up (what does it set up? idk?? the project). i added a line about `draw()` doing motion

- changed "canvas" to "the HTML canvas element"

- not gonna go into detail about rgb. it is not necessary for this. i'll just add a note "play around with the values" and ugh why is there no webpage that just says
"what is rgb?" "it's a way to define colors for the web by specifying red, green, and blue values"
i dont want to add that into the workshop cuz this workshop not about random stuff like this. if there were a column for footnotes this is where i would put it.
also i already linked to the bg documentation. should i just link every function to docuemntation? eh i dont see the harm. i'll do this

### add player

- added some fluff about `var player;` but idk if it's adequate

- added desc about width and height

- "placeholder for image": idk what the sprite is. a customizable object? idk????!?!

- ug idk how to plop this coordinate system in the beginning. "oh here's a library p5js oh btw the coordinate system is different"
i need to rethink the introduction of p5js / canvas. i don't think it's clear enough that we're working with like, a section of the page that we can create pictures/visuals on.

- annoyed that the color of the sprite randomly changes
this is annoying and sometimes it changes to a really dark color that blends in with the bg

- no @ height-40. the code tells u that it's not being cut off.

### add motion

- fuck ur wording

- earlier i changed `draw()` a bit so maybe that helps

- unclear where to write code: lollll ok changed

### add controls

- linking todocumentation 5eva

- idfk @ conditionals but this workshop should def not be the workshop to introduce ALL the concepts it covers cuz that is a lot of concepts
behold, are you understanding why i wanted to break up workshops by concept now. uh huh.
that said, i think soundboard is gonna be dope for introing functions

- it is not unclear. i stated "let's wrap the prev line in this conditional"

### keeping player in bounds

- wording is disgusting in this section but WHATEV

### enemy

### moving enemy

- random() -fixed

### game over

- hmm idk maybe i can create the empty function body in that section?

- flag : yea idk im iffy on using this. got any suggestions?

- idk if explaining `return` is easier than explaining `else { }`

### restart

### error handling

- i want to rename this "covering our asses"

### spicing it up

### code cleanup

- idk i just wanted to stick w the setting up being in setup(). i guess if i initialized the constants when declaring, i could use this to point out the diff btwn variables and constants? "the reason we can do this here is that it's never gonna change"
but idk idk. why do we not declare the player and enemy also? that also never 'changes,' from being "createSprite(dfsfsdfs)" idk

jkwok91 commented on workshops/dodge/README.md:
```diff
@@ -2,54 +2,696 @@
+
+p5.js has a function called [`loadImage()`](http://p5js.org/reference/#p5/loadImage) that takes URL of an image as an argument and gives us a loaded image ready to be used. p5.play sprites have the method [`addImage()`](http://p5play.molleindustria.org/docs/classes/Sprite.html#method-addImage) that we can give a loaded image to assign it to the sprite.
+
+In addition to `setup()` and `draw()`, p5.play has a special function called `preload()`, which is run right when the page loads before `setup()`. `preload()` is generally used to load images and other resources (like sounds) into the game. Go ahead and declare `preload()` right before the `setup()` function.
```
er, p5.js, not p5.play

jkwok91 commented on workshops/dodge/README.md:
```diff
@@ -2,54 +2,696 @@
+
+I've chosen this dark blue, but you can experiment with the RGB values and choose your own background color.
+
+Now that we've finished setting up, it's time to add the most important thing to our game: you, the player.
```
yea i'm gonna veto this "you" addition because i refer to "you" as "the user" throughout the workshop, and explicitly avoid conflating "the player," "you," and "the user."

jkwok91 commented on workshops/dodge/README.md:
```diff
@@ -2,54 +2,696 @@
+function setup() {
+  ...
+  p̶l̶a̶y̶e̶r̶ ̶=̶ ̶c̶r̶e̶a̶t̶e̶S̶p̶r̶i̶t̶e̶(̶w̶i̶d̶t̶h̶/̶2̶,̶ ̶h̶e̶i̶g̶h̶t̶-̶2̶5̶,̶ ̶5̶0̶,̶ ̶5̶0̶)̶;̶
+  player = createSprite(width/2, height-25, 0, 0);
```
I guess this is better than telling people that you don't have to include those arguments at all? Hah.

zachlatta replied:
Yeah, I figure that it's clearer to not introduce the concept of loosely
typed arguments yet.

On Fri, Apr 8, 2016 at 8:14 PM, JESSICA <notifications@github.com> wrote:

> In workshops/dodge/README.md
> <https://github.com/hackclub/hackclub/pull/781#discussion_r59109398>:
>
> > +```js
> > +function setup() {
> > +  ...
> > +  player = createSprite(width/2, height-25, 50, 50);
> > +  player.addImage(playerImage);
> > +  ...
> > +}
> > +```
> > +
> > +When we add an image to a sprite, the sprite's width and height is changed to match the image's. Because we're loading our sprite's image right after we create the sprite, we don't need to worry about the sprite's initial width and height. Let's just go ahead and set them to zero.
> > +
> > +```js
> > +function setup() {
> > +  ...
> > +  p̶l̶a̶y̶e̶r̶ ̶=̶ ̶c̶r̶e̶a̶t̶e̶S̶p̶r̶i̶t̶e̶(̶w̶i̶d̶t̶h̶/̶2̶,̶ ̶h̶e̶i̶g̶h̶t̶-̶2̶5̶,̶ ̶5̶0̶,̶ ̶5̶0̶)̶;̶
> > +  player = createSprite(width/2, height-25, 0, 0);
>
> I guess this is better than telling people that you don't have to include
> those arguments at all? Hah.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/hackclub/hackclub/pull/781/files/bb7e11407098f5cf6ca44a9e996b10f9884834b0#r59109398>
>



-- 
Zach Latta
https://zachlatta.com <http://zachlatta.com>

zachlatta commented:
![](http://i.giphy.com/OjBocHTklyQda.gif)
